ckanext-datagov_inventory/ckanext/datagov_inventory/plugin.py

```diff
@@ -0,0 +1,213 @@
+import ckan.plugins as plugins
+import ckan.plugins.toolkit as toolkit
+import ckan.authz as authz
+import logging
+
+
+log = logging.getLogger(__name__)
+
+
+# a generic user validation function
+default_message = 'You must login to perform this action.'
+
+
+def validate_user(context, data_dict=None):
+    user = context.get('user')
+    if user:
+        return {'success': True}
```

Comment:
Based on the docs, it doesn't look like CKAN will chain auth calls by default. This looks to me like as long as you are a user, you're authorized, ignoring any additional auth checks in default CKAN. Maybe it doesn't matter since we're only looking at read actions where you're authorized or not, but if you threw this on a group-specific action it wouldn't end well.

I feel like what we really want is to implement this as a decorator around the existing behavior. The docs on [chained_auth_function](https://docs.ckan.org/en/2.8/extensions/plugins-toolkit.html#ckan.plugins.toolkit.ckan.plugins.toolkit.chained_auth_function) are a bit light, with [more detail here](https://docs.ckan.org/en/2.8/extensions/plugin-interfaces.html#ckan.plugins.interfaces.IAuthFunctions.get_auth_functions), and an example: https://github.com/vrk-kpa/api-catalog/blob/27551a8a90f51fdc6dcbba0e6c4418a51cc291bc/ckanext/ckanext-apicatalog_routes/ckanext/apicatalog_routes/auth.py#L48-L54

I think we're really after:

```python
# Implemented as a decorator (I may be a bit rusty here on the syntax)
def datagov_disallow_anonymous_access(func=None):
  def _wrapper(next_auth, context, dict_data=None):
    if func:
      # We're called in the decorator context, call the function
      return func(next_auth, context, dict_data)
    else:
      # We were called as a function, just call the next_auth function
      return next_auth(context, dict_data)

  # We're always applying the auth_disallow_anonymous_access decorator.
  return ckan.plugin.toolkit.auth_disallow_anonymous_access(_wrapper)

# Used as a decorator for custom auth logic, if necessary
@datagov_disallow_anonymous_access
def datagov_group_list_authz(next_auth, context, dict_data):
  # custom logic maybe return
  return next_auth(context, dict_data)

# In most cases, don't define a custom auth function at all, just return our chained wrapper
# which is already decorated with auth_disallow_anonymous_access
class Plugin:
    def get_auth_functions(self):
        return {
                'format_autocomplete': datagov_disallow_anonymous_access(),
                'group_list': datagov_disallow_anonymous_access(),
                'group_list_authz': datagov_group_list_authz,
                # ...
        }
```

Reply:
That decorator code is probably wrong, I think I'm missing the chained_auth_function decorator. It's probably more like:

```python
def datagov_disallow_anonymous_access(func=None):
  @chained_auth_function
  @auth_disallow_anonymous_access
  def _wrapper(next_auth, context, dict_data=None):
    # ...

  return _wrapper
```

Reply:
The above is correct. There's an open CKAN bug and and a PR for this. [Chained auth functions lose anonymous access #5751](https://github.com/ckan/ckan/issues/5751).

Reply:
Fwiw I would not wait for the bug fix.  It's easy enough to include the user check in our own decorator.